pallets/asset-index/src/lib.rs

```diff
@@ -673,6 +713,51 @@ pub mod pallet {
 			share.checked_div(&nav).ok_or_else(|| ArithmeticError::Overflow.into())
 		}
 
+		/// Consolidate and clean deposits while completing withdraw
+		fn do_consolidate_deposits(caller: &T::AccountId, mut amount: T::Balance) -> Result<T::Balance, DispatchError> {
+			let withdrawal_amount = amount.clone();
+			<Deposits<T>>::try_mutate_exists(&caller, |maybe_depositing| -> Result<T::Balance, DispatchError> {
+				let depositing = maybe_depositing.take().ok_or(<Error<T>>::NoDeposits)?;
+				let mut start: T::BlockNumber = 0_u32.into();
+
+				let still_depositing: Vec<_> = depositing
+					.into_iter()
+					.filter_map(|(mut index_tokens, block_number)| {
+						// check depositing duration
+						if block_number < start || start == 0_u32.into() {
+							start = block_number;
+						}
+
+						// consolidate deposits
+						if block_number > frame_system::Pallet::<T>::block_number() {
```

Comment:
checking the block number here doesn't make sense here
we need to subtract the `index_tokens` from `amount`, 
if `amount` > `index_tokens` we delete the deposit entry and add `totalfee+=redemption_fee(index_tokens, block_number)`
else if `index_tokens` > `amount` we update value and `totalfee+=redemption_fee(index_tokens -amount, block_number)`